### ISSUE TITLE
Failure when trying to run pod create

### ISSUE THREAD

**Issue by chavafg** [01org/cc-oci-runtime#815]
There seems to be a random issue when trying to launch containers.
As example when running this test: 
```
not ok 27 Kill container with SIGUSR1
# (in test file /home/cc-tester/go/src/github.com/01org/cc-oci-runtime/cc-oci-runtime-2.1.3/tests/integration/docker/kill.bats, line 55)
#   `$DOCKER_EXE ps -a | grep $container | grep "Exited (${exit_code})"' failed
# 9a1dc3f385305255200413d47bd57242591203eb383e864aced317883a9ab712
# Mg51LRvmYTRN1nIXiTQFsY60cw7qUMO4
```

Looking at the logs, cc-oci-runtime fails to run `pod create `
```
2017-04-11T13:17:57.678926Z:21266:cc-oci-runtime:debug:CTL created event: 1
2017-04-11T13:17:57.696571Z:21019:cc-oci-runtime:debug:proxy msg length: 37
2017-04-11T13:17:57.696662Z:21019:cc-oci-runtime:debug:message read from proxy socket: {"success":true,"data":{"version":1}}
2017-04-11T13:17:57.696781Z:21019:cc-oci-runtime:debug:msg received: {"success":true,"data":{"version":1}}
2017-04-11T13:17:57.696957Z:21019:cc-oci-runtime:debug:communicating with proxy
2017-04-11T13:17:57.697004Z:21019:cc-oci-runtime:debug:sending message (length 598) to proxy socket
2017-04-11T13:17:57.697049Z:21019:cc-oci-runtime:debug:writing message data to proxy socket: {"id":"hyper","data":{"hyperName":"startpod","data":{"hostname":"11c29395d10c","containers":[],"shareDir":"rootfs","in
terfaces":[{"macAddr":"02:42:0a:ff:00:07","newDeviceName":"eth0","mtu":1450,"ipAddresses":[{"ipAddress":"10.255.0.7","netMask":"255.255.0.0"},{"ipAddress":"10.255.0.4","netMask":"255.255.255.255"}]},{"macAddr":"
02:42:ac:12:00:03","newDeviceName":"eth1","mtu":1500,"ipAddresses":[{"ipAddress":"172.18.0.3","netMask":"255.255.0.0"}]}],"routes":[{"dest":"default","gateway":"172.18.0.1","device":"eth1"},{"dest":"10.255.0.0",
"device":"eth0"},{"dest":"172.18.0.0","device":"eth1"}]}}}
2017-04-11T13:17:57.724485Z:21019:cc-oci-runtime:debug:proxy msg length: 58
2017-04-11T13:17:57.724580Z:21019:cc-oci-runtime:debug:message read from proxy socket: {"success":false,"error":"ERROR received from Hyperstart"}
2017-04-11T13:17:57.724672Z:21019:cc-oci-runtime:critical:failed to run hyper cmd startpod: {"success":false,"error":"ERROR received from Hyperstart"}
2017-04-11T13:17:57.724746Z:21019:cc-oci-runtime:critical:failed to run pod create
2017-04-11T13:17:57.724845Z:21183:cc-oci-runtime:critical:failed to read proxy socket fd

```

**Comment by devimc** [comment]
@sboeuf @amshinde @mcastelino 
seems like MTU failed

```
I0411 14:30:00.432457    8734 proxy.go:77] [client #31] hyper(cmd=startpod, data={"hostname":"3470d8ea98c0","containers":[],"shareDir":"rootfs","interfaces":[{"macAddr":"02:42:ac:11:00:02","newDeviceName":"eth0","mtu":1500,"ipAddresses":[{"ipAddress":"172.17.0.2","netMask":"255.255.0.0"}]}],"routes":[{"dest":"default","gateway":"172.17.0.1","device":"eth0"},{"dest":"172.17.0.0","device":"eth0"}]})
I0411 14:30:00.432556    8734 vm.go:114] [vm 3470d8ea hyperstart] hyper_loop epoll_wait 1
I0411 14:30:00.432586    8734 vm.go:114] [vm 3470d8ea hyperstart] hyper_handle_event get event 1, he 0x4201f0, fd 3. ops 0x420030
I0411 14:30:00.432609    8734 vm.go:114] [vm 3470d8ea hyperstart] hyper_handle_event event EPOLLIN, he 0x4201f0, fd 3, 0x420030
I0411 14:30:00.432641    8734 vm.go:114] [vm 3470d8ea hyperstart] hyper_channel_read
I0411 14:30:00.432663    8734 vm.go:114] [vm 3470d8ea hyperstart] hyper send type 14, len 4
I0411 14:30:00.432693    8734 vm.go:114] [vm 3470d8ea hyperstart] get length 326
I0411 14:30:00.432716    8734 vm.go:114] [vm 3470d8ea hyperstart] hyper send type 14, len 4
I0411 14:30:00.432798    8734 vm.go:114] [vm 3470d8ea hyperstart] 0 0 0 1 0 0 1 46 7b 22 68 6f 73 74 6e 61 6d 65 22 3a 22 33 34 37 30 64 38 65 61 39 38 63 30 22 2c 22 63 6f 6e 74 61 69 6e 65 72 73 22 3a 5b 5d 2c 22 73 68 61 72 65 44 69 72 22 3a 22 72 6f 6f 74 66 73 22 2c 22 69 6e 74 65 72 66 61 63 65 73 22 3a 5b 7b 22 6d 61 63 41 64 64 72 22 3a 22 30 32 3a 34 32 3a 61 63 3a 31 31 3a 30 30 3a 30 32 22 2c 22 6e 65 77 44 65 76 69 63 65 4e 61 6d 65 22 3a 22 65 74 68 30 22 2c 22 6d 74 75 22 3a 31 35 30 30 2c 22 69 70 41 64 64 72 65 73 73 65 73 22 3a 5b 7b 22 69 70 41 64 64 72 65 73 73 22 3a 22 31 37 32 2e 31 37 2e 30 2e 32 22 2c 22 6e 65 74 4d 61 73 6b 22 3a 22 32 35 35 2e 32 35 35 2e 30 2e 30 22 7d 5d 7d 5d 2c 22 72 6f 75 74 65 73 22 3a 5b 7b 22 64 65 73 74 22 3a 22 64 65 66 61 75 6c 74 22 2c 22 67 61 74 65 77 61 79 22 3a 22 31 37 32 2e 31 37 2e 30 2e 31 22 2c 22 64 65 76 69 63 65 22 3a 22 65 74 68 30 22 7d 2c 7b 22 64 65 73 74 22 3a 22 31 37 32 2e 31 37 2e 30 2e 30 22 2c 22 64 65 76 69 63 65 22 3a 22 65 74 68 30 22 7d 5d 7d 
I0411 14:30:00.432813    8734 vm.go:114] [vm 3470d8ea hyperstart]  hyper_channel_handle, type 1, len 326
I0411 14:30:00.432851    8734 vm.go:114] [vm 3470d8ea hyperstart] call hyper_start_pod, json {"hostname":"3470d8ea98c0","containers":[],"shareDir":"rootfs","interfaces":[{"macAddr":"02:42:ac:11:00:02","newDeviceName":"eth0","mtu":1500,"ipAddresses":[{"ipAddress":"172.17.0.2","netMask":"255.255.0.0"}]}],"routes":[{"dest":"default","gateway":"172.17.0.1","device":"eth0"},{"dest":"172.17.0.0","device":"eth0"}]}, len 318
I0411 14:30:00.432953    8734 vm.go:114] [vm 3470d8ea hyperstart] call hyper_start_pod, json {"hostname":"3470d8ea98c0","containers":[],"shareDir":"rootfs","interfaces":[{"macAddr":"02:42:ac:11:00:02","newDeviceName":"eth0","mtu":1500,"ipAddresses":[{"ipAddress":"172.17.0.2","netMask":"255.255.0.0"}]}],"routes":[{"dest":"default","gateway":"172.17.0.1","device":"eth0"},{"dest":"172.17.0.0","device":"eth0"}]}, len 318
I0411 14:30:00.433001    8734 vm.go:114] [vm 3470d8ea hyperstart] jsmn parse successed, n is 37
I0411 14:30:00.433066    8734 vm.go:114] [vm 3470d8ea hyperstart] token 0, type is 1, size is 5
I0411 14:30:00.433090    8734 vm.go:114] [vm 3470d8ea hyperstart] token 1, type is 3, size is 1
I0411 14:30:00.433133    8734 vm.go:114] [vm 3470d8ea hyperstart] hostname is 3470d8ea98c0
I0411 14:30:00.433155    8734 vm.go:114] [vm 3470d8ea hyperstart] token 3, type is 3, size is 1
I0411 14:30:00.433176    8734 vm.go:114] [vm 3470d8ea hyperstart] exec hwclock -s command failed: No such file or directory
I0411 14:30:00.433202    8734 vm.go:114] [vm 3470d8ea hyperstart] container count 0
I0411 14:30:00.433220    8734 vm.go:114] [vm 3470d8ea hyperstart] token 5, type is 3, size is 1
I0411 14:30:00.433242    8734 vm.go:114] [vm 3470d8ea hyperstart] share tag is rootfs
I0411 14:30:00.433264    8734 vm.go:114] [vm 3470d8ea hyperstart] token 7, type is 3, size is 1
I0411 14:30:00.433287    8734 vm.go:114] [vm 3470d8ea hyperstart] network interfaces num 1
I0411 14:30:00.433318    8734 vm.go:114] [vm 3470d8ea hyperstart] net mac address is 02:42:ac:11:00:02
I0411 14:30:00.433341    8734 vm.go:114] [vm 3470d8ea hyperstart] new interface name is eth0
I0411 14:30:00.433364    8734 vm.go:114] [vm 3470d8ea hyperstart] mtu is 1500
I0411 14:30:00.433386    8734 vm.go:114] [vm 3470d8ea hyperstart] Number of ip addresses:1
I0411 14:30:00.433408    8734 vm.go:114] [vm 3470d8ea hyperstart] net ipaddress for device (null) is 172.17.0.2
I0411 14:30:00.433435    8734 vm.go:114] [vm 3470d8ea hyperstart] net mask for device (null) is 255.255.0.0
I0411 14:30:00.433470    8734 vm.go:114] [vm 3470d8ea hyperstart] token 23, type is 3, size is 1
I0411 14:30:00.433493    8734 vm.go:114] [vm 3470d8ea hyperstart] network routes num 2
I0411 14:30:00.433518    8734 vm.go:114] [vm 3470d8ea hyperstart] route 0 dest is default
I0411 14:30:00.433537    8734 vm.go:114] [vm 3470d8ea hyperstart] route 0 gateway is 172.17.0.1
I0411 14:30:00.433561    8734 vm.go:114] [vm 3470d8ea hyperstart] route 0 device is eth0
I0411 14:30:00.433581    8734 vm.go:114] [vm 3470d8ea hyperstart] route 1 dest is 172.17.0.0
I0411 14:30:00.433604    8734 vm.go:114] [vm 3470d8ea hyperstart] route 1 device is eth0
I0411 14:30:00.433640    8734 vm.go:114] [vm 3470d8ea hyperstart] create directory /tmp/hyper
I0411 14:30:00.433700    8734 vm.go:114] [vm 3470d8ea hyperstart] create directory /tmp/hyper/proc
I0411 14:30:00.433832    8734 vm.go:114] [vm 3470d8ea hyperstart] nic eth0 already exists
I0411 14:30:00.433855    8734 vm.go:114] [vm 3470d8ea hyperstart] net device eth0
I0411 14:30:00.433877    8734 vm.go:114] [vm 3470d8ea hyperstart] net device sys path is /sys/class/net/eth0/ifindex
I0411 14:30:00.433918    8734 vm.go:114] [vm 3470d8ea hyperstart] get ifindex 2
I0411 14:30:00.433949    8734 vm.go:114] [vm 3470d8ea hyperstart] interface get netamsk 16 255.255.0.0
I0411 14:30:00.434049    8734 vm.go:114] [vm 3470d8ea hyperstart] Setting interface MTU to 1500
I0411 14:30:00.434379    8734 vm.go:114] [vm 3470d8ea hyperstart] net device eth0
I0411 14:30:00.434402    8734 vm.go:114] [vm 3470d8ea hyperstart] net device sys path is /sys/class/net/eth0/ifindex
I0411 14:30:00.434470    8734 vm.go:114] [vm 3470d8ea hyperstart] can not open file: No such file or directory
I0411 14:30:00.434651    8734 vm.go:114] [vm 3470d8ea hyperstart] failed to get the ifindix of eth0
I0411 14:30:00.434672    8734 vm.go:114] [vm 3470d8ea hyperstart] setup route failed
I0411 14:30:00.434704    8734 vm.go:114] [vm 3470d8ea hyperstart] setup network failed
I0411 14:30:00.434727    8734 vm.go:114] [vm 3470d8ea hyperstart] hyper send type 10, len 0
I0411 14:30:00.434847    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /proc/sys/fs/binfmt_misc
I0411 14:30:00.434917    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/fs/cgroup/devices
I0411 14:30:00.435014    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/fs/cgroup/cpu,cpuacct
I0411 14:30:00.435112    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/fs/cgroup/perf_event
I0411 14:30:00.435164    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/fs/cgroup/freezer
I0411 14:30:00.435216    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/fs/cgroup/memory
I0411 14:30:00.435287    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/fs/cgroup/net_cls
I0411 14:30:00.435340    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/kernel/debug
I0411 14:30:00.435382    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/fs/fuse/connections
I0411 14:30:00.435409    8734 proxy.go:67] [client #31] connection closed
I0411 14:30:00.435425    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /dev/mqueue
I0411 14:30:00.435520    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /sys/fs/cgroup/systemd
I0411 14:30:00.435560    8734 vm.go:114] [vm 3470d8ea hyperstart] umount /
```

**Comment by sboeuf** [comment]
@devimc Looking at those traces, MTU seems to be set properly. Maybe an issue with the routes.

**Comment by jodh-intel** [comment]
Relates to new test introduced on #807.

**Comment by amshinde** [comment]
From the logs, looks like there was a failure in getting the index of the eth0 device the second time :
"failed to get the ifindix of eth0".
Will take a look at why this is happening.

**Comment by amshinde** [comment]
We introduced a change in hyperstart to get the interface name based on the mac address(hyper_get_iface_name_from_mac_addr). But we do this before the wait_for_nic(ifname) call we had implemented earlier, essentially causing to not wait at all. So, sometimes the  earlier function returns "eth0" address instead of the usual pci based address ("enp0s4").

I added an explicit call to wait_for_nic("enp0s4"), with that I see this udev event:
```
move@/devices/pci0000:00/0000:00:04.0/virtio2/net/enp0s4^M
ACTION=move^M
DEVPATH=/devices/pci0000:00/0000:00:04.0/virtio2/net/enp0s4^M
SUBSYSTEM=net^M
DEVPATH_OLD=/devices/pci0000:00/0000:00:04.0/virtio2/net/eth0^M
INTERFACE=enp0s4^M
IFINDEX=2^M
SEQNUM=659^M
ACTION=move^M
```

A subsequent call to match the mac address based on interface name gives "ioctl SIOCGIFHWADDR failed: No such device"
So it appears there is lag where the kernel is naming eth0 to enps04 according the the above event even after the event is received in user space, and we try to match the interface name with the mac address before the rename is complete.

I was thinking of changing the wait based on mac address instead of interface name, but I dont see that passed in the udev event.
Also, with the above observed lag between the event and the rename, I am not sure if we can rely on the event at all. I propose to add a change to fetch interface index based on mac address and use the index throughout instead of relying on the interface name at all.
